pkg/metallb/frr_test.go

```diff
@@ -0,0 +1,321 @@
+
+		if testCase.expectedError == "" {
+			assert.Equal(t, testCase.name, testFRRConfigurationBuilder.Definition.Name)
+			assert.Equal(t, testCase.namespace, testFRRConfigurationBuilder.Definition.Namespace)
+		}
+	}
+}
+
+func TestFrrConfigurationWithBGPPassword(t *testing.T) {
+	testCases := []struct {
+		testFrrConfiguration *FrrConfigurationBuilder
+		bgpPassword          string
+		neighborInt          int
+		expectedError        string
+	}{
+		{
+			testFrrConfiguration: buildValidFRRConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
+			bgpPassword:          "bgpPassword",
+			expectedError:        "",
+		},
+		{
+			testFrrConfiguration: buildValidFRRConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
+			bgpPassword:          "",
+			expectedError:        "the bgpPassword  is an empty string",
+		},
+	}
+
+	for _, testCase := range testCases {
+		frrConfigurationBuilder := testCase.testFrrConfiguration.WithBGPRouter(64550).
+			WithBGPNeighbor("10.46.73.131", 64500).
+			WithBGPPassword(testCase.bgpPassword, 0)
+		assert.Equal(t, testCase.expectedError, frrConfigurationBuilder.errorMsg)
+
+		if testCase.expectedError == "" {
+			assert.Equal(t, testCase.bgpPassword,
+				frrConfigurationBuilder.Definition.Spec.BGP.Routers[0].Neighbors[0].Password)
+		}
+	}
+}
+
+func TestFrrConfigurationWithHoldTime(t *testing.T) {
+	testCases := []struct {
+		testFrrConfiguration *FrrConfigurationBuilder
+		holdTime             *metav1.Duration
+		expectedError        string
+	}{
+		{
+			testFrrConfiguration: buildValidFRRConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
+			holdTime: &metav1.Duration{
+				Duration: 90 * time.Second,
+			},
+		},
+		{
+			testFrrConfiguration: buildValidFRRConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
+			holdTime: &metav1.Duration{
+				Duration: 0 * time.Minute,
+			},
+			expectedError: "frrConfiguration 'holdtime' value is not valid",
+		},
+	}
+
+	for _, testCase := range testCases {
+		frrConfigurationBuilder := testCase.testFrrConfiguration.WithBGPRouter(64550).
+			WithBGPNeighbor("10.46.73.131", 64500).WithHoldTime(*testCase.holdTime, 0)
+		assert.Equal(t, testCase.expectedError, frrConfigurationBuilder.errorMsg)
+
+		if testCase.expectedError == "" {
+			assert.Equal(t, testCase.holdTime,
+				frrConfigurationBuilder.Definition.Spec.BGP.Routers[0].Neighbors[0].HoldTime)
+		}
+	}
+}
+
+func TestFrrConfigurationWithKeepalive(t *testing.T) {
+	testCases := []struct {
+		testFrrConfiguration *FrrConfigurationBuilder
+		keepalive            *metav1.Duration
+		expectedError        string
+	}{
+		{
+			testFrrConfiguration: buildValidFRRConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
+			keepalive: &metav1.Duration{
+				Duration: 30 * time.Second,
+			},
+			expectedError: "",
+		},
+		{
+			testFrrConfiguration: buildInValidFrrConfigurationBuilder(buildFrrConfigurationTestClientWithDummyObject()),
```

Comment:
I think should be buildValidFRRConfigurationBuilder not buildInValidFrrConfigurationBuilder

Reply:
yes. changed